### PR TITLE
Add support for `torch.gather`

### LINKED ISSUE #223: Add support for torch.gather
## 🚀 Feature

`torch.gather` may be used for GNNs and is required to implement backward of `torch.scatter_add`.

cc @apaz-cli

### PR DESCRIPTION
**Pull request by rdspring1**
## What does this PR do?

This PR implements gather prim, adds support for `torch.gather`, and implements its grad transform.

Fixes #223.

### CODE REVIEW

File: thunder/tests/opinfos.py [hunk 1]
```diff
@@ -4345,6 +4345,57 @@ def take_along_axis_sample_generator(op, device, dtype, requires_grad, **kwargs)
...
+    torch_reference=torch.gather,
+    test_directives=(
+        # Torch doesn't support complex half on gather
+        DecorateInfo(
+            pytest.mark.skip,
+            "test_core_vs_torch_consistency",
+            dtypes=(datatypes.complex32,),
+        ),
+        DecorateInfo(
+            custom_comparator(partial(assert_close, atol=1e-1, rtol=1e-1)),
```
Comment (IvanYashchuk) [inline thunder/tests/opinfos.py:4389]: The tolerances are a bit too high it seems, what's the failure rate with something smaller?

File: thunder/tests/opinfos.py [hunk 2]
```diff
@@ -4345,6 +4345,57 @@ def take_along_axis_sample_generator(op, device, dtype, requires_grad, **kwargs)
...
+        canonicalized_dim = dim if dim >= 0 else dim + len(shape_a)
+        a = make(shape_a)
+        b = make_index(shape_b, low=0, high=shape_a[dim])
+        yield SampleInput(a, index=b, dim=dim)
+
+    # Questionable use case. Do we want to support these?!
+    # Note that gather doesn't have the broadcast requirement, it only requires
+    # 1. a.shape[i]      >= index.shape[i] for i != dim
+    #
+    # a.shape, dim, index.shape
```
Comment (IvanYashchuk) [inline thunder/tests/opinfos.py:4364]: Does nvFuser support this?

**Review by IvanYashchuk — APPROVED** [review]

File: thunder/tests/opinfos.py (hunk 1 above)
Comment (nikitaved) [reply thunder/tests/opinfos.py:4389]: Yeah, this is strange. Gather just copies, so if this tolerance is needed it is likely that something isn't right..

File: thunder/tests/opinfos.py (hunk 2 above)
Comment (nikitaved) [reply thunder/tests/opinfos.py:4364]: Questionable in what sense? Seems totally fine for me, and it allows to really select a subset of things.

File: thunder/tests/opinfos.py (hunk 1 above)
Comment (nikitaved) [reply thunder/tests/opinfos.py:4389]: Ah, well, I guess the issue is backward? Well, the indices gotta be unique. Only then the impl backward is correct.

Comment (rdspring1) [reply thunder/tests/opinfos.py]: I remove the test_directives; The tests pass.

Comment (nikitaved) [reply thunder/tests/opinfos.py]: @rdspring1 , it is still weird, as the backward formula assumes no repeated indices. Otherwise it is incorrect. Do we create such index tensors by chance? I am not sure we fix the seed just yet, so, maybe, it is best to resolve that right now?

**Review by nikitaved — APPROVED** [review]
LGTM!

**Comment by IvanYashchuk** [comment]
Code owner review required @carmocca, @t-vi.

**Review by carmocca — APPROVED** [review]

File: thunder/core/transforms.py [hunk 3]
```diff
@@ -815,6 +815,23 @@ def _take_prim_grad(a: TensorProxy, index: TensorProxy, dim: int) -> TensorProxy
 register_grad(pids.TAKE, _take_prim_grad)
 
 
+@torchctx
+def _gather_prim_grad(a: TensorProxy, index: TensorProxy, dim: int) -> TensorProxy:
+    fwd = prims.gather(a, index, dim)
+
+    g = get_grad(fwd)
+    # NOTE Intentionally not calling zeros_like to avoid preserving a
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply thunder/core/transforms.py]: `a` in this context is a variable name for a `TensorProxy`. I updated the comment but it was a complete sentence.
